server/aesdsocket.c

```diff
@@ -78,40 +80,57 @@ void signal_handler(int signal) {
 void *connection_handler(void *socket_desc) {
     int connfd = *(int *)socket_desc;
     char *buffer = calloc(BUFFER_SIZE, sizeof(char));
+    int fd = open(fdir, O_RDWR); //Open the device file only once and use the same fd for IOCTL and reads
     
+    if (fd < 0) {
+        perror("open: Failed to open device.");
+        free(buffer);
+        close(connfd);
+        return NULL;
+    }
+
     while (recv(connfd, buffer, BUFFER_SIZE, 0) > 0) {
```

Comment:
@tiba6275 since you are assuming null terminated strings you need the `len` to be `BUFFER_SIZE -1` here.